convert_zdf.py

```diff
@@ -0,0 +1,150 @@
+"""Convert ZDF point cloud to the available formats.
+
+Example: $ python -u convert_zdf.py --ply Zivid3D.zdf
+
+Available formats:
+    .ply - Polygon File Format
+    .csv,.txt - [X, Y, Z, r, g, b, Contrast]
+    .png,.jpg,.bmp,.tiff - 2D RGB image
+"""
+
+import argparse
+from pathlib import Path
+import numpy as np
+import cv2
+import zivid
+
+
+def _options():
+    parser = argparse.ArgumentParser(
+        description="Convert from ZDF to your preferred format\
+            \nExample:\n\t $ python -u convert_zdf.py --ply Zivid3D.zdf",
+        formatter_class=argparse.RawTextHelpFormatter,
+    )
+
+    parser.add_argument("filename", help="File to convert", default="Zivid3D.zdf")
+    group = parser.add_mutually_exclusive_group(required=True)
+    group.add_argument("--ply", action="store_true", help="Convert to PLY")
+    group.add_argument("--csv", action="store_true", help="Convert to CSV")
+    group.add_argument("--txt", action="store_true", help="Convert to text")
+    group.add_argument(
+        "--png",
+        action="store_true",
+        help="Convert to Portable Network Graphics (3D->2D)",
+    )
+    group.add_argument("--jpg", action="store_true", help="Convert to JPEG (3D->2D)")
+    group.add_argument(
+        "--bmp", action="store_true", help="Convert to Windows bitmap (3D->2D)"
+    )
+    group.add_argument("--tiff", action="store_true", help="Convert to TIFF (3D->2D)")
+
+    return parser.parse_args()
+
+
+def _flatten_point_cloud(point_cloud):
+    """Convert from point cloud to flattened point cloud (with numpy).
+
+    Args:
+        point_cloud: a numpy array
+
+    Returns:
+        2D numpy array, with 7 columns and npixels rows
+
+    """
+    # Convert to numpy 3D array
+    point_cloud = np.dstack(
+        [
+            point_cloud["x"],
+            point_cloud["y"],
+            point_cloud["z"],
+            point_cloud["r"],
+            point_cloud["g"],
+            point_cloud["b"],
+            point_cloud["contrast"],
+        ]
+    )
+    # Flattening the point cloud
+    flattened_point_cloud = point_cloud.reshape(-1, 7)
+    # Just the points without color and contrast
+    # point_cloud = np.dstack([point_cloud['x'],point_cloud['y'],point_cloud['z']])
+    # flattened_point_cloud = point_cloud.reshape(-1,3)
+
+    # Removing nans
+    return flattened_point_cloud[~np.isnan(flattened_point_cloud[:, 0]), :]
+
+
+def _convert_2_ply(frame, file_name: str):
+    """Convert from frame to ply.
+
+    Args:
+        frame: A frame captured by a Zivid camera.
+        file_name: File name without extension
+
+    """
+    print(f"Saving the frame to {file_name}.ply")
+    frame.save(file_name + ".ply")
```

Comment:
I would have use an f-string here actually

Reply:
Just out of curiousity, why? file_name is already a string what's then the benefit of f"{file_name}.ply"? Is the latter more explicit and thus portable?

Reply:
`string1 + string2` is the old way of concatenating strings.
It usually becomes much more readable if you use a f-string:
f"{string2}{string2}"
This is a trivial example, so everything here should be more or less readable. But consider this string:
`f"useradd -m -u {uid} -g {gid} {username}"`
vs
`"useradd -m -u " + uid + " -g " + gid + " " + username"`

Reply:
I agree with that, and in general I do use f-string

Reply:
In this case specifically:
`f"{file_name}.ply"` is clearly more obvious compared to `file_name + ".ply"`

Reply:
This is a great article about the different formatting possibilities:
https://realpython.com/python-f-strings/

Reply:
I believe I read this, or a very similar one, a year or more ago.